File: be/Controllers/HallController.cs
using FA23_Convocation2023_API.DTO;
using FA23_Convocation2023_API.Services;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;

namespace FA23_Convocation2023_API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class HallController : ControllerBase
    {
        private readonly HallService _hallService;

        public HallController(HallService hallService)
        {
            _hallService = hallService;
        }

        //Create a new hall
        [HttpPost("CreateHall")]
        public async Task<IActionResult> CreateHallAsync([FromBody]CreateHallRequest hallRequest)
        {
            var hallExist = await _hallService.HallExist(hallRequest.HallName);
            if (hallExist)
            {
                return BadRequest(new
                {
                    status = StatusCodes.Status400BadRequest,
                    message = "Hall already exists!"
                });
            }
            
            var result = await _hallService.CreateHall(hallRequest.HallName);
            return Ok(new
            {
                status = StatusCodes.Status200OK,
                message = "Create hall successfully!",
                data = result
            });
        }

        //get list hall
        [HttpGet("GetAll")]
        public async Task<IActionResult> GetAllHallAsync()
        {
            var result = await _hallService.GetAllHallAsync();
            if (result.Count == 0) return Ok(new
            {
                status = StatusCodes.Status204NoContent,
                message = "Not any halls!"
            });
            return Ok(new
            {
                status = StatusCodes.Status200OK,
                message = "Get all halls successfully!",
                data = result
            });
        }

        [HttpPut("UpdateHall/{hallId}")] // Update hall
        public async Task<IActionResult> UpdateHallAsync([FromRoute] int hallId, [FromBody] string hallName)
        {
            var result = await _hallService.UpdateHallAsync(hallId, hallName);
            if (!result) return BadRequest(new
            {
                status = StatusCodes.Status400BadRequest,
                message = "Update failed!"
            });
            return Ok(new
            {
                status = StatusCodes.Status200OK,
                message = "Update hall successfully!",
                data = result
            });
        }

        [HttpDelete("DeleteHall/{hallId}")] // Delete hall
        public async Task<IActionResult> DeleteHallAsync([FromRoute] int hallId)
        {
            var result = await _hallService.DeleteHallAsync(hallId);
            if (!result) return BadRequest(new
            {
                status = StatusCodes.Status400BadRequest,
                message = "Xóa thất bại! Hãy chắc chắn trong Hall không còn Bachelor, Checkins nào"
            });
            return Ok(new
            {
                status = StatusCodes.Status200OK,
                message = "Delete hall successfully!",
                data = result
            });
        }
    }
}

File: be-nest/src/hall/hall.controller.ts
import { Body, Controller, Delete, Get, Param, ParseIntPipe, Post, Put } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Permission } from '../auth/permissions.js';
import { Public } from '../common/guards/public.decorator.js';
import { RequirePermissions } from '../common/guards/require-permissions.decorator.js';
import { CreateHallDto } from './dto/hall.dto.js';
import { HallService } from './hall.service.js';

@ApiTags('hall')
@Controller('Hall')
export class HallController {
  constructor(private readonly halls: HallService) {}

  @Post('CreateHall')
  @RequirePermissions(Permission.ManageHalls)
  async create(@Body() input: CreateHallDto) {
    return {
      status: 200,
      message: 'Tạo hội trường thành công!',
      data: await this.halls.create(input.hallName),
    };
  }

  @Get('GetAll')
  @Public()
  async list() {
    const data = await this.halls.list();
    return data.length
      ? { status: 200, message: 'Lấy danh sách hội trường thành công!', data }
      : { status: 204, message: 'Không có hội trường nào!' };
  }

  @Put('UpdateHall/:hallId')
  @RequirePermissions(Permission.ManageHalls)
  async update(@Param('hallId', ParseIntPipe) id: number, @Body() name: string) {
    const success = await this.halls.update(id, name.trim());
    if (!success) return { status: 400, message: 'Cập nhật hội trường thất bại!' };
    return { status: 200, message: 'Cập nhật hội trường thành công!', data: true };
  }

  @Delete('DeleteHall/:hallId')
  @RequirePermissions(Permission.ManageHalls)
  async delete(@Param('hallId', ParseIntPipe) id: number) {
    const success = await this.halls.delete(id);
    if (!success) {
      return {
        status: 400,
        message: 'Xóa thất bại! Hãy chắc chắn hội trường không còn tân cử nhân hoặc bản ghi check-in nào.',
      };
    }
    return { status: 200, message: 'Xóa hội trường thành công!', data: true };
  }
}
